File: bak/apps/api/src/modules/admin/admin-tasks.controller.ts
import {
  Controller,
  Get,
  Inject,
  Param,
  Query,
  Res,
  StreamableFile,
  UseGuards,
} from "@nestjs/common";
import { GenerationResultAssetsService } from "../generation/generation-result-assets.service";
import { AdminPermissionGuard, RequireAdminPermission } from "./admin-permission.guard";
import { AdminTasksService } from "./admin-tasks.service";

@Controller("admin/tasks")
@UseGuards(AdminPermissionGuard)
@RequireAdminPermission("tasks:read")
export class AdminTasksController {
  constructor(
    @Inject(AdminTasksService) private readonly service: AdminTasksService,
    @Inject(GenerationResultAssetsService) private readonly assets: GenerationResultAssetsService,
  ) {}

  @Get()
  list(
    @Query("status") status?: string,
    @Query("model") model?: string,
    @Query("query") query?: string,
    @Query("createdFrom") createdFrom?: string,
    @Query("createdTo") createdTo?: string,
    @Query("page") page?: string,
    @Query("pageSize") pageSize?: string,
  ) {
    return this.service.list({ status, model, query, createdFrom, createdTo, page, pageSize });
  }

  @Get("results/:resultId/content")
  async readResult(
    @Param("resultId") resultId: string,
    @Res({ passthrough: true }) response: AssetResponse,
  ) {
    return this.serveAsset(await this.assets.readAny(resultId, "content"), response);
  }

  @Get("results/:resultId/thumbnail")
  async readResultThumbnail(
    @Param("resultId") resultId: string,
    @Res({ passthrough: true }) response: AssetResponse,
  ) {
    return this.serveAsset(await this.assets.readAny(resultId, "thumbnail"), response);
  }

  @Get("reconciliation/runs")
  listReconciliationRuns() {
    return this.service.listReconciliationRuns();
  }

  @Get(":taskId")
  get(@Param("taskId") taskId: string) {
    return this.service.get(taskId);
  }

  private serveAsset(
    asset: { redirectUrl: string | null; data: Buffer | null; mimeType: string },
    response: AssetResponse,
  ) {
    if (asset.redirectUrl) {
      response.redirect(302, asset.redirectUrl);
      return;
    }
    if (!asset.data) return;
    response.setHeader("Content-Type", asset.mimeType);
    response.setHeader("Content-Length", String(asset.data.byteLength));
    response.setHeader("Content-Disposition", 'inline; filename="generation-result.webp"');
    response.setHeader("Cache-Control", "private, max-age=3600");
    response.setHeader("X-Content-Type-Options", "nosniff");
    return new StreamableFile(asset.data);
  }
}

interface AssetResponse {
  redirect(statusCode: number, url: string): void;
  setHeader(name: string, value: string): void;
}

File: dream_service/api/src/main/java/com/dreamspace/api/controller/AdminTasksController.java
package com.dreamspace.api.controller;

import com.dreamspace.api.common.AdminPermission;
import com.dreamspace.api.service.AdminTasksService;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/manage_web/tasks")
@AdminPermission
public class AdminTasksController {
  private final AdminTasksService service;

  public AdminTasksController(AdminTasksService service) { this.service = service; }

  @GetMapping
  AdminTasksService.Page list(@RequestParam(required = false) String status,
      @RequestParam(required = false) String model, @RequestParam(required = false) String query,
      @RequestParam(required = false) String createdFrom,
      @RequestParam(required = false) String createdTo,
      @RequestParam(defaultValue = "1") int page,
      @RequestParam(defaultValue = "20") int pageSize) {
    return service.list(status, model, query, createdFrom, createdTo, page, pageSize);
  }

  @GetMapping("/results/{resultId}/content")
  ResponseEntity<byte[]> content(@PathVariable String resultId) { return binary(resultId, false); }

  @GetMapping("/results/{resultId}/thumbnail")
  ResponseEntity<byte[]> thumbnail(@PathVariable String resultId) { return binary(resultId, true); }

  @GetMapping("/reconciliation/runs")
  AdminTasksService.ReconciliationResponse reconciliation() { return service.reconciliation(); }

  @GetMapping("/{taskId}")
  AdminTasksService.TaskDetail detail(@PathVariable String taskId) { return service.get(taskId); }

  private ResponseEntity<byte[]> binary(String resultId, boolean thumbnail) {
    var data = service.readResult(resultId, thumbnail);
    String extension = data.contentType().toLowerCase().endsWith("png") ? "png" : "webp";
    return ResponseEntity.ok().contentType(MediaType.parseMediaType(data.contentType()))
        .contentLength(data.bytes().length)
        .header(HttpHeaders.CONTENT_DISPOSITION, "inline; filename=\"generation-result." + extension + "\"")
        .header(HttpHeaders.CACHE_CONTROL, "private, max-age=3600")
        .header("X-Content-Type-Options", "nosniff").body(data.bytes());
  }
}
